Remove old frontend and ui backend. Update Readme. (#73)
neuralet/smart-social-distancing: README.md

cd smart-social-distancing
```

### Download a sample video
If you don't have any camera to test the solution you can use any video as an input source. You can download an example with the following command.

```bash
# Download a sample video file from multiview object tracking dataset
# The video is complied from this dataset: https://researchdatafinder.qut.edu.au/display/n27416
./download_sample_video.sh
```

### Usage
The smart social distancing app consists of two components: the `frontend` and the `processor`. 

#### Frontend

> The frontend is a public [webapp](http://beta.lanthorn.ai) provided by [lanthorn](https://www.lanthorn.ai/) where you can signup for free. This web app allows you to configure some aspects of the processor (such as notifications and calibration) using a friendly UI. Moreover, it provides a dashboard that helps you to analyze the data that your cameras are processing. 

Reviewer: We normally prefer to have each sentence on a new line in markdown files so that PR comments are easier to process.

```suggestion
The frontend is a public [web app](http://beta.lanthorn.ai) provided by [lanthorn](https://www.lanthorn.ai/) where you can signup for free. 
This web app allows you to configure some aspects of the processor (such as notifications and camera calibration) using a friendly UI. 
Moreover, it provides a dashboard that helps you to analyze the data that your cameras are processing. 
```

**Persisting changes on the config.ini file**

> Adding the respective `config.ini` as a volume (as for example `-v "$PWD/config-jetson.ini":/repo/config-jetson.ini`). Will allow for syncing changes of said file. 

Reviewer: Please discard this change

Author: Done

`config-x86-openvino.ini`: for x86 systems accelerated with Openvino

You can also modify some of them using the [UI](https://beta.lanthorn.ai). 

> If you choose this option, take care of mount the config file to keep the changes after any restart of the processor.

Reviewer: ```suggestion
If you choose this option, make sure to mount the config file as a volume to keep the changes after any restart of the container.
```

All the configurations are grouped in *sections* and some of them can vary depending on the chosen device.

- `[App]`
  - `Resolution`: Specifies the image resolution that the whole processor will use. If you are using a single camera we recommend using that resolution.
  - `Encoder`: Specifies the video encoder used by the processing pipeline.
  - `MaxProcesses`: Defines the number of processes executed in the processor. If you are using multiple cameras per processor we recommend increasing this number.

>   - `ScreenshotPeriod`: Defines a time period (expressed in minutes) to take a screenshot of all the cameras ans store them in S3. If you set the value in 0 no screenshots are taken.

Reviewer: ```suggestion
  - `ScreenshotPeriod`: Defines a time period (expressed in minutes) to take a screenshot of all the cameras and store them in S3. If you set the value to 0, no screenshots will be taken.
```

  - `ScreenshotS3Bucket`: Configures the S3 Bucket used to store the screenshot.

>   - `ScreenshotsDirectory`: Configures the folder dedicated to storing all the images (for example the heatmap reports) generated by the processor. We recommend set this folder in a mounted directory (such as */repo/data/processor/static/screenshots*)

Reviewer: ```suggestion
  - `ScreenshotsDirectory`: Configures the folder dedicated to storing all the images (for example the heatmap reports) generated by the processor. We recommend to set this folder to a mounted directory (such as */repo/data/processor/static/screenshots*)
```

Is this directory used for heatmaps as the text says?

  - `DashboardURL`: Sets the url where the frontend is running. Unless you are using a custom domain, you should keep this value as https://beta.lanthorn.ai/.

>   - `EnableSlackNotifications`: A boolean parameter to enable/disable the Slack integration for notifications and daily reports. We recommend not editing this parameter directly and manage it from the [UI](https://beta.lanthorn.ai) to configure your workspace correclty.

Reviewer: ```suggestion
  - `EnableSlackNotifications`: A boolean parameter to enable/disable the Slack integration for notifications and daily reports. We recommend not editing this parameter directly and manage it from the [UI](https://beta.lanthorn.ai) to configure your workspace correctly.
```

>   - `SlackChannel`: Configures the slack channel used by the notifications. The slack channel chosen must exist in the workspace configured.

Reviewer: ```suggestion
  - `SlackChannel`: Configures the slack channel used by the notifications. The chosen slack channel must exist in the configured workspace.
```

- `[Api]`

>   - `Host`: Configures the host IP of the processor's API (inside docker). We recommend don't change that value and keep it as *0.0.0.0*.

Reviewer: ```suggestion
  - `Host`: Configures the host IP of the processor's API (inside docker). We recommend not to change that value and keep it as *0.0.0.0*.
```

  - `Post`: Configures the port of the processor's API (inside docker). Take care that if you change the default value (*8000*) you will need to change the startup command to expose the configured endpoint.

>   - `SSLEnabled`: A boolean parameter to enable/disable https/ssl in the API. We recommend setting this value in *True*. 

Reviewer: ```suggestion
  - `SSLEnabled`: A boolean parameter to enable/disable HTTPS/SSL in the API. We recommend setting this value to *True*. 
```

  - `SSLCertificateFile`: Specifies the location of the SSL certificate (required when you have *SSL enabled*). If you generate it following the steps defined in this Readme you should put */repo/certs/<your_ip>.crt*
  - [`SSLKeyFile`]: Specifies the location of the SSL key file (required when you have *SSL enabled*). If you generate it following the steps defined in this Readme you should put */repo/certs/<your_ip>.crt*

- `[Core]`:
  - `Host`: Sets the host IP of the *QueueManager* (inside docker).
  - `QueuePort`: Sets the port of the *QueueManager* (inside docker).
  - `QueueAuthKey`: Configures the auth key required to interact with the *QueueManager*.

- `[Area_N]`:

  A single processor can manage multiple areas and all of them must be configured in the config file. You can generate this configuration in 3 different ways: directly in the config file, using the [UI](https://beta.lanthorn.ai) or using the API.
  - `Id`: A string parameter to identify each area. This value must be *unique*.
  - `Name`: A string parameter to name each area. Although you can repeat the same name in multiple areas, we recommend don't do that.

>   - `Cameras`: Configures the cameras (using the *ids*) included in the area. If you are configuring multiple cameras you should write the ids separated by commas. Each area should have unless one camera.

Reviewer: ```suggestion
  - `Cameras`: Configures the cameras (using the *ids*) included in the area. If you are configuring multiple cameras you should write the ids separated by commas. Each area should have at least one camera.
```

>   - `NotifyEveryMinutes` and `ViolationThreshold`: Defines the *period of time* and *number of social distancing violations* desired to send notifications. For example, if you want to notify when *occurs more than 10 violations every 15 minutes*, you must set `NotifyEveryMinutes` in 15 and `ViolationThreshold` in 10.

Reviewer: ```suggestion
  - `NotifyEveryMinutes` and `ViolationThreshold`: Defines the *period of time* and *number of social distancing violations* desired to send notifications. For example, if you want to notify when *more than 10 violations happen in 15 minutes*, you must set `NotifyEveryMinutes` in 15 and `ViolationThreshold` in 10.
```

We wanted to move away from the `violations` term and use `close contact` instead. But I think social distancing violations is clear

>   -`Emails`: Defines the emails list to receive the notification. Multiple emails can be written separating them by commas.

Reviewer: ```suggestion
  - `Emails`: Defines the email list to receive the notifications. Multiple emails can be written separating them by commas.
```

>   - `OccupancyThreshold`: Defines the occupancy violation threshold. For example, if you want to notify when *there is more than 20 persons in the area* you must set `OccupancyThreshold` in 20.

Reviewer: ```suggestion
  - `OccupancyThreshold`: Defines the occupancy violation threshold. For example, if you want to notify when *there are more than 20 persons in the area* you must set `OccupancyThreshold` to 20.
```

>   - `DailyReport`: When the parameter is set in *True*, the information of the previous day is sent in a summary report.

Reviewer: ```suggestion
  - `DailyReport`: When the parameter is set to *True*, the information of the previous day is sent in a summary report.
```

  - `DailyReportTime`: If the daily report is enabled, you can choose the time to receive the report. By default, the report is sent at 06:00.

- `[Source_N]`:

>   In the config files, we use the *source* sections to specifies the camera's configurations. Similarly to the areas, a single processor can manage multiple cameras and all of them must be configured in the config file. You can generate this configuration in 3 different ways: directly in the config file, using the [UI](https://beta.lanthorn.ai) or using the API.

Reviewer: ```suggestion
  In the config files, we use the *source* sections to specify the  configurations of the camera/video feeds to process. 
  Similarly to the areas, a single processor can manage multiple cameras and all of them must be configured in the config file. 
  You can generate this configuration in 3 different ways: directly in the config file, using the [UI](https://beta.lanthorn.ai) or using the API.
```

  - `Id`: A string parameter to identify each camera. This value must be *unique*.
  - `Name`: A string parameter to name each area. Although you can repeat the same name in multiple cameras, we recommend don't do that.
  - `VideoPath`: Sets the path or url required to get the camera's video stream.

>   - `Tags`: List of tags (separated by commas). This field only has an informative propose, change that value doesn't affect the processor behavior.

Reviewer: ```suggestion
  - `Tags`: List of tags (separated by commas). This field only has an informative purpose, changing that value doesn't affect the processor behavior.
```

>   - `NotifyEveryMinutes` and `ViolationThreshold`: Defines the *period of time* and *number of social distancing violations* desired to send notifications. For example, if you want to notify when *occurs more than 10 violations every 15 minutes*, you must set `NotifyEveryMinutes` in 15 and `ViolationThreshold` in 10.

Reviewer: ```suggestion
  - `NotifyEveryMinutes` and `ViolationThreshold`: Defines the *period of time* and *number of social distancing violations* desired to send notifications. For example, if you want to notify when *more than 10 violations happen in 15 minutes*, you must set `NotifyEveryMinutes` to 15 and `ViolationThreshold` to 10.
```

>   -`Emails`: Defines the emails list to receive the notification. Multiple emails can be written separating them by commas.

Reviewer: ```suggestion
  - `Emails`: Defines the email list to receive the notifications. Multiple emails can be written separating them by commas.
```

>   - `DailyReport`: When the parameter is set in *True*, the information of the previous day is sent in a summary report.

Reviewer: ```suggestion
  - `DailyReport`: When the parameter is set to *True*, the information of the previous day is sent in a summary report.
```

  - `DailyReportTime`: If the daily report is enabled, you can choose the time to receive the report. By default, the report is sent at 06:00.
  - `DistMethod`: Configures the chosen distance method used by the processor to detect the violations. There are three different values: CalibratedDistance, CenterPointsDistance and FourCornerPointsDistance. If you want to use *CalibratedDistance* you will need to calibrate the camera from the [UI](https://beta.lanthorn.ai).

- `[Detector]`:
  - `Device`: Specifies the device. The available values are *Jetson*, *EdgeTPU*, *Dummy*, *x86*
  - `Name`: Defines the detector's models used by the processor. The models available varies from device to device. Information about the supported models are specified in a comment in the corresponding *config-<device>.ini* file.

>   - `ImageSize`: Configures the moedel input size. When the image has a different resolution, it is resized to fit the model ones. The available values of this parameter depends on the model chosen.

Reviewer: ```suggestion
  - `ImageSize`: Configures the model's input size. When the image has a different resolution, it is resized to fit the model's requirements. The available values for this parameter depends on the chosen model.
```

  - `ModelPath`: Some of the supported models allow you to overwrite the default one. For example, if you have a specific model trained for your scenario you can use it.
  - `ClassID`: When you are using a multi-class detection model, you can definde the class id related to pedestrian in this parameter.
  - `MinScore`: Defines the person detection threshold. Any person detected by the model with a score less than the threshold will be ignored.

- `[Classifier]`:

  Some of the supported devices includes the *facemask detection* feature. If you want to include this feature, you need to specify the classifier section.
  - `Name`: Name of the facemask classifier used.

>   - `ImageSize`: Configures the moedel input size. When the image has a different resolution, it is resized to fit the model ones. The available values of this parameter depends on the model chosen.

Reviewer: ```suggestion
  - `ImageSize`: Configures the model input size. When the image has a different resolution, it is resized to fit the model's requirements. The available values of this parameter depend on the chosen model.
```

>   - `ModelPath`: The same behavior that the section `Detector`.

Reviewer: ```suggestion
  - `ModelPath`: The same behavior as in the section `Detector`.
```

- `[PostProcessor]`:
  - `MaxTrackFrame`: Defines the number of frames that an object should disappear to be considered as lost.

>   - `NMSThreshold`: Configures the threshold of minimum IoU of to detect two box as duplicated.

Reviewer: ```suggestion
  - `NMSThreshold`: Configures the threshold of minimum IoU to detect two boxes as referring to the same object.
```

  - `DefaultDistMethod`: Defines the default distance algorithm for the cameras without *DistMethod* configuration.
  - `DistThreshold`: Configures the distance threshold for the *social distancing violations*
  - `Anonymize`: A boolean parameter to enable/disable anonymization of faces in videos and screenshots.

- `[Logger]`:
  - `Name`: Configures the chosen logger. For now, we only support the *csv_logger*.
  - `TimeInterval`: Sets the desired logging interval for objects detections and violations.
  - `LogDirectory`: Defines the location where the generated files will be stored.
  - `EnableReports`: A boolean parameter to enable/disable the reports generation.
  - `HeatmapResolution`: Sets the resolution used by the heatmap report.

### API usage
After you run the processor on your node, you can use the exposed API to control the Processor's Core, where all the process is getting done.

The available endpoints are grouped in the following subapis:
- `/config`: provides a pair of endpoint to retrieve and overwrite the current configuration file.

> - `/cameras`: provides endpoints to execute all the CRUD operations required by cameras. This endpoint sare very useful to edit the camera's configuration without restarting the docker process. Additionally, this subapi exposes some calibration endpoints.

Reviewer: ```suggestion
- `/cameras`: provides endpoints to execute all the CRUD operations required by cameras. These endpoints are very useful to edit the camera's configuration without restarting the docker process. Additionally, this subapi exposes some calibration endpoints.
```

- `/areas`: provides endpoints to execute all the CRUD operations required by areas.
- `/reports`: a set of endpoints to retrieve the data generated by the processor.

> - `/slack`: a set of endpoints required to configure slack correctly in the processor. We recommend use these endpoints from the [UI](https://beta.lanthorn.ai)instead of calling them directly.

Reviewer: ```suggestion
- `/slack`: a set of endpoints required to configure Slack correctly in the processor. We recommend to use these endpoints from the [UI](https://beta.lanthorn.ai) instead of calling them directly.
```

>  Additionally, the api exposes 2 endpointa to stop/start the video processing

Reviewer: ```suggestion
 Additionally, the API exposes 2 endpoints to stop/start the video processing
```
